lib/semantics/check-do.cc

```diff
@@ -64,6 +64,76 @@ class DoConcurrentBodyEnforce {
   template<typename T> bool Pre(const T &) { return true; }
   template<typename T> void Post(const T &) {}
 
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+
+  // C1140 -- Can't deallocate a polymorphic entity in a DO CONCURRENT.
+  // Deallocation can be caused by exiting a block that declares an allocatable
+  // entity, assignment to an allocatable variable, or an actual DEALLOCATE
+  // statement
+  //
+  // Note also that the deallocation of a derived type entity might cause the
+  // invocation of an IMPURE final subroutine.
+
+  // Deallocation caused by block exit
+  void Post(const parser::BlockConstruct &blockConstruct) {
+    const auto &endBlockStmt{
+        std::get<parser::Statement<parser::EndBlockStmt>>(blockConstruct.t)};
+    const Scope &blockScope{context_.FindScope(endBlockStmt.source)};
+    const Scope &doScope{context_.FindScope(doConcurrentSourcePosition_)};
+    if (DoesScopeContain(&doScope, blockScope)) {
+      for (auto &pair : blockScope) {
+        Symbol &symbol{*pair.second};
+        if (IsOrContainsPolymorphicComponent(symbol) && IsAllocatable(symbol) &&
```

Comment:
Shouldn't the allocatable check be for the polymorphic component?

Reply:
Wow.  Thanks for pointing this out.  I hadn't realized that the ```ALLOCATABLE``` attribute could be associated with a component.

So deallocation of a polymorphic entity can happen when you have something that's allocatable that either is or contains a polymorphic entity.  I'll redo my checking code to test for this more complex condition and add tests to cover the relevant cases.